chain/client/src/stateless_validation/partial_witness_tracker.rs

```diff
@@ -0,0 +1,168 @@
+use std::collections::HashMap;
+
+use lru::LruCache;
+use near_chain::Error;
+use near_primitives::reed_solomon::rs_decode;
+use near_primitives::sharding::ChunkHash;
+use near_primitives::stateless_validation::{EncodedChunkStateWitness, PartialEncodedStateWitness};
+use reed_solomon_erasure::galois_8::ReedSolomon;
+
+/// Max number of chunks to keep in the witness tracker cache. We reach here only after validation
+/// of the partial_witness so the LRU cache size need not be too large.
+const NUM_CHUNKS_IN_WITNESS_TRACKER_CACHE: usize = 200;
+
+/// Ratio of the number of data parts to total parts in the Reed Solomon encoding.
+/// The tradeoff here is having a higher ratio is better for handling missing parts and network errors
+/// but increases the size of the encoded state witness and the total network bandwidth requirements.
+const RATIO_DATA_PARTS: f32 = 0.8;
+
+/// Reed Solomon encoder for encoding state witness parts.
+/// We keep one wrapper for each length of chunk_validators to avoid re-creating the encoder.
+/// This is used by `PartialEncodedStateWitnessTracker`
+pub struct RsMap {
```

Comment:
I suggest to keep it simple and have have `ReedSolomon` instance per `CacheEntry`, I guess we can afford that

Reply:
I think that would also solve [this problem](https://github.com/near/nearcore/pull/11170/files#r1582679541) where I wasn't handling the case of having a single part only.